monitoring/check_collectd_mlab_test.py

```diff
@@ -0,0 +1,301 @@
+    self.assertRaises(
+        check_collectd_mlab.SocketConnectionError,
+        check_collectd_mlab.sock_connect, 'no_socket_name')
+
+  def testunit_sock_readline_WHEN_socket_error_RAISES_Error(self):
+    mock_sock = mock.Mock(spec_set=socket.socket)
+    mock_sock.recv.side_effect = socket.error('fake error')
+
+    self.assertRaises(
+        check_collectd_mlab.SocketReadlineError,
+        check_collectd_mlab.sock_readline, mock_sock)
+
+
+class MLabCollectdAssertionTests(unittest.TestCase):
+
+  def setUp(self):
+    self._testdata_dir = os.path.join(
+        os.path.dirname(check_collectd_mlab.__file__), 'testdata')
+    check_collectd_mlab.COLLECTD_BIN = (
+        os.path.join(self._testdata_dir, 'fake_bin'))
+    check_collectd_mlab.COLLECTD_NAGIOS = (
+        os.path.join(self._testdata_dir, 'fake_bin'))
+    check_collectd_mlab.COLLECTD_PID = (
+        os.path.join(self._testdata_dir, 'fake_pid'))
+    check_collectd_mlab.COLLECTD_UNIXSOCK = (
+        os.path.join(self._testdata_dir, 'fake_socket'))
+    check_collectd_mlab.VSYSPATH_BACKEND = (
+        os.path.join(self._testdata_dir, 'fake_backend'))
+    check_collectd_mlab.VSYSPATH_SLICE = (
+        os.path.join(self._testdata_dir, 'fake_slice'))
+
+  @mock.patch('check_collectd_mlab.sock_connect')
+  def testunit_assert_collectd_responds_WHEN_sock_sendcmd_fails(
+      self, mock_sock_connect):
+    mock_sock = mock.Mock(spec_set=socket.socket)
+    mock_sock.send.side_effect = socket.error('fake socket error')
+    mock_sock_connect.return_value = mock_sock
+
+    self.assertRaises(
+        check_collectd_mlab.SocketSendCommandError,
+        check_collectd_mlab.assert_collectd_responds)
+    mock_sock_connect.assert_called_with(
+        check_collectd_mlab.COLLECTD_UNIXSOCK)
+
+  @mock.patch('check_collectd_mlab.sock_connect')
+  def testunit_assert_collectd_responds_WHEN_sock_readline_fails(
+      self, mock_sock_connect):
+    mock_sock = mock.Mock(spec_set=socket.socket)
+    # After sending a default reply, send EOF ('') prematurely.
+    mock_sock.recv.side_effect = list('1 default reply\n') + ['']
+    mock_sock_connect.return_value = mock_sock
+
+    self.assertRaises(
+        check_collectd_mlab.SocketReadlineError,
+        check_collectd_mlab.assert_collectd_responds)
+    mock_sock_connect.assert_called_with(
+        check_collectd_mlab.COLLECTD_UNIXSOCK)
+
+  @mock.patch('os.access')
+  def testunit_assert_collectd_responds_WHEN_filesystem_is_readonly(
+      self, mock_access):
+    # Mock out os.access to guarantee that write access is rejected.
+    mock_access.return_value = False
+
+    self.assertRaises(
+        check_collectd_mlab.ReadonlyFilesystemError,
+        check_collectd_mlab.assert_collectd_responds)
+    self.assertTrue(mock_access.called)
+
+  def testunit_assert_collectd_responds_WHEN_sock_connect_fails(self):
+    check_collectd_mlab.COLLECTD_UNIXSOCK = 'does_not_exist'
```

Comment:
Ahhhh okay. I was getting lost because there's a few layers of indirection here in mocking out the sock vs. mocking out the return of sock.connect.

One thing I've been trying to do recently when I feel like the test function name isn't clear enough is writing additional explanation in a one-line docstring, because one-line docstrings get printed to the console when the test fails along with the function name. So in this case I'd suggest:

```
  def testunit_assert_collectd_responds_WHEN_sock_sendcmd_fails(
      self, mock_sock_connect):
  """Fail when we create a successful connection, but hit an error on send."""
...

  def testunit_assert_collectd_responds_WHEN_sock_connect_fails(self):
  """Fail if creating a socket connection fails."""
```